mobile_verifier/src/promotion_reward.rs

```diff
@@ -0,0 +1,565 @@
+    promotion_rewards_sink: FileSinkClient<VerifiedPromotionRewardV1>,
+}
+
+impl PromotionRewardDaemon {
+    pub async fn create_managed_task(
+        pool: PgPool,
+        settings: &Settings,
+        file_upload: FileUpload,
+        file_store: file_store::FileStore,
+        carrier_client: CarrierServiceClient,
+    ) -> anyhow::Result<impl ManagedTask> {
+        let (promotion_rewards_sink, valid_promotion_rewards_server) =
+            VerifiedPromotionRewardV1::file_sink(
+                settings.store_base_path(),
+                file_upload.clone(),
+                Some(Duration::from_secs(15 * 60)),
+                env!("CARGO_PKG_NAME"),
+            )
+            .await?;
+
+        let (promotion_rewards, promotion_rewards_server) =
+            file_source::continuous_source::<PromotionReward, _>()
+                .state(pool.clone())
+                .store(file_store)
+                .lookback(LookbackBehavior::StartAfter(settings.start_after))
+                .prefix(FileType::PromotionRewardIngestReport.to_string())
+                .create()
+                .await?;
+
+        let promotion_reward_daemon = Self {
+            pool,
+            carrier_client,
+            promotion_rewards,
+            promotion_rewards_sink,
+        };
+
+        Ok(TaskManager::builder()
+            .add_task(valid_promotion_rewards_server)
+            .add_task(promotion_rewards_server)
+            .add_task(promotion_reward_daemon)
+            .build())
+    }
+
+    async fn run(mut self, shutdown: triggered::Listener) -> anyhow::Result<()> {
+        loop {
+            tokio::select! {
+                _ = shutdown.clone() => {
+                    tracing::info!("PromotionRewardDaemon shutting down");
+                    break;
+                }
+                Some(file) = self.promotion_rewards.recv() => {
+                    let start = Instant::now();
+                    self.process_file(file).await?;
+                    metrics::histogram!("promotion_reward_processing_time")
+                        .record(start.elapsed());
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn process_file(&self, file: FileInfoStream<PromotionReward>) -> anyhow::Result<()> {
+        tracing::info!("Processing promotion reward file {}", file.file_info.key);
+
+        let mut transaction = self.pool.begin().await?;
+        let reports = file.into_stream(&mut transaction).await?;
+
+        let mut verified_promotion_rewards = pin!(
+            VerifiedPromotionReward::validate_promotion_rewards(reports, &self.carrier_client,)
+        );
+
+        while let Some(promotion_reward) = verified_promotion_rewards.try_next().await? {
+            promotion_reward.write(&self.promotion_rewards_sink).await?;
+            if promotion_reward.is_valid() {
+                promotion_reward.save(&mut transaction).await?;
+            }
+        }
```

Comment:
I'm struggling with "verified" and "validate" in this context. 

If we're getting Verified Promotions by validating promotions it seems weird to ask if `promo.is_valid()` when iterating them.

Maybe `VerifiedPromotionReward::from_promotion_rewards(reports, &self.carrier_client)`? Then reading this usage, it doesn't seem redundant to call `promo.is_valid()` before saving.

Reply:
I renamed the struct to "ValidatedPromotionReward" to be more in line with everything else